### PR TITLE
fix(lib-storage): add missing return keys

### LINKED ISSUE #2694: lib-storage Upload working but result is missing
## Describe the bug
Hello there,

Since 3.26.0 the `Upload` command from **_lib-storage_** does not return the same data and therefore is not usable

Edit: to clarify : the file is sent correctly to the bucket, but the result is not correct

#### SDK version number
```
@aws-sdk/client-s3: 3.26.0 and 3.27.0
@aws-sdk/lib-storage: 3.26.0 and 3.27.0
```

#### Is the issue in the browser/Node.js/ReactNative?
Node.js

#### Details of the browser/Node.js/ReactNative version
```
▶ node -v
v16.2.0
```

## Steps to reproduce
```typescript
const client = new S3({
    region,
    credentials: {
        accessKeyId,
        secretAccessKey,
     },
 })
const params = {
    Bucket: bucket,
    Key: filename,
    Body: file,
    ACL: 'public-read',
    ContentType: type,
}

const upload = new Upload({
    client,
    queueSize: 3,
    params,
})
const data = await upload.done()

console.log({ data })
```

#### Observed behavior
With 3.26.0 and 3.27.0, the content of data is
```typescript
{
  data: {
    '$metadata': {
      httpStatusCode: 200,
      requestId: undefined,
      extendedRequestId: 'dD4fCNw0SfXPpz6AKXc......',
      cfId: undefined,
      attempts: 1,
      totalRetryDelay: 0
    },
    BucketKeyEnabled: undefined,
    ETag: '"8e79af4e71fa......."',
    Expiration: undefined,
    RequestCharged: undefined,
    SSECustomerAlgorithm: undefined,
    SSECustomerKeyMD5: undefined,
    SSEKMSEncryptionContext: undefined,
    SSEKMSKeyId: undefined,
    ServerSideEncryption: 'AES256',
    VersionId: 'OCh1_......'
  }
}
```

#### Expected behavior
with 3.25.0 the result is 
```typescript
{
  data: {
    '$metadata': {
      httpStatusCode: 200,
      requestId: undefined,
      extendedRequestId: 'H2cT3+QuQ3vG2/fBtj/GP8g...',
      cfId: undefined,
      attempts: 1,
      totalRetryDelay: 0
    },
    Bucket: 'The name of the bucket',
    BucketKeyEnabled: undefined,
    ETag: '"5a101d2..."',
    Expiration: undefined,
    Key: 'the key provided',
    Location: 'https://xxx.s3.eu-central-1.amazonaws.com/the key provided',
    RequestCharged: undefined,
    SSEKMSKeyId: undefined,
    ServerSideEncryption: 'AES256',
    VersionId: '28mF6wpfmGAe...'
  }
}
```

### PR DESCRIPTION
**Pull request by samchungy**
### Issue
Resolves #2694
### Description
- Add `Bucket`, `Key` and `Location` fields to the return output of lib-storage when PUT is used.

### Testing
How was this change tested?
- Unit Tests
- Compared Output of 3.25 with 3.27.

### Additional context
I took the format of the location key from a v3.25 Upload and used it to form the logic for the string creation. `Location: 'https://kickoff-dev-serverlessdeploymentbucket-fzcuxfqyzlsl.s3.ap-southeast-2.amazonaws.com/test'`, 

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/lib-storage/src/Upload.ts [hunk 1]
```diff
@@ -212,7 +213,25 @@ export class Upload extends EventEmitter {
...
+      const locationBucket = extendedEncodeURIComponent(this.params.Bucket!);
+
+      let Location: string;
+      if (this.client.config.forcePathStyle) {
+        Location = `${endpoint.protocol}//${endpoint.hostname}/${locationBucket}/${locationKey}`;
+      } else {
+        Location = `${endpoint.protocol}//${locationBucket}.${endpoint.hostname}/${locationKey}`;
+      }
+      result = {
+        ...this.putResponse!,
```
Comment (Linkgoron) [inline lib/lib-storage/src/Upload.ts:230]: Wouldn't it make more sense to add it directly to the object (although you would probably need to change some types)?

**Comment by wallind** [comment]
any idea if this is slotted to make its way into a release at some point? Not to be impatient just wondering since the last activity was 26 days ago

File: lib/lib-storage/src/Upload.ts [hunk 2]
```diff
@@ -96,8 +97,29 @@ export class Upload extends EventEmitter {
...
+
+    const locationKey = this.params
+      .Key!.split("/")
+      .map((segment) => extendedEncodeURIComponent(segment))
+      .join("/");
+    const locationBucket = extendedEncodeURIComponent(this.params.Bucket!);
+
+    let Location: string;
+    if (this.client.config.forcePathStyle) {
+      Location = `${endpoint.protocol}//${endpoint.hostname}/${locationBucket}/${locationKey}`;
```
Comment (samchungy) [inline lib/lib-storage/src/Upload.ts:113]: I did some playing around with endpoint.path -> but couldn't quite figure out it's behavior, not sure if this needs adjusting.

**Comment by samchungy** [comment]
> any idea if this is slotted to make its way into a release at some point? Not to be impatient just wondering since the last activity was 26 days ago

Sorry - no clue 🤷‍♂️  I'm just a contributor not a maintainer

File: lib/lib-storage/src/Upload.ts [hunk 3]
```diff
@@ -96,8 +97,29 @@ export class Upload extends EventEmitter {
...
+    } else {
+      Location = `${endpoint.protocol}//${locationBucket}.${endpoint.hostname}/${locationKey}`;
+    }
```
Comment (trivikr) [inline lib/lib-storage/src/Upload.ts:116]: Use const
```suggestion
    const Location = this.client.config.forcePathStyle
      ? `${endpoint.protocol}//${endpoint.hostname}/${locationBucket}/${locationKey}`
      : `${endpoint.protocol}//${locationBucket}.${endpoint.hostname}/${locationKey}`;
```

**Comment by wallind** [comment]
> Sorry - no clue 🤷‍♂️ I'm just a contributor not a maintainer

fair enough haha. Exciting to see some movement on this. If you end up needing a hand hit me back and maybe I'll take a look at this. I'm tired of having to skip the Renovate PRs around this dependency 🙃

**Comment by Narretz** [comment]
@Linkgoron @ajredniwja can this PR get some attention again? It's actually embarassing that this migration option for the upload has been broken for so long.

**Comment by Linkgoron** [comment]
> @Linkgoron @ajredniwja can this PR get some attention again? It's actually embarassing that this migration option for the upload has been broken for so long.

I'm not a part of the AWS team...

**Comment by wallind** [comment]
> @Linkgoron @ajredniwja can this PR get some attention again? It's actually embarassing that this migration option for the upload has been broken for so long.

I wouldn't have phrased it this way but in sentiment I agree. Is there anything I can do to help expedite this as I'd really lock to remove the version lock I have to get around it.

Is it just that we're waiting on maintainers to come back through?

**Comment by Linkgoron** [comment]
> Would it be possible to approve this PR? This is really annoying behavior... @Linkgoron @trivikr @ajredniwja

@AlbanDurrheimer As I've already stated elsewhere, please don't ping me... I'm not a part of the team (nor was I ever). Sadly, it was my contribution that caused this bug, which is why I commented in the issue and in the PR, but otherwise I an unrelated to AWS.

**Comment by ryanto** [comment]
Thanks for fixing this!

I maintain a wrapper library around S3 (next-s3-upload) that would benefit a ton from lib-storage. However, the library depends on the `Location` return from AWS. From my understand it looks like I wouldn't be able to access `Location` until this PR is merged.

Is there anything I can do to help get this PR merged? Or, in the mean time, is there any advice for accessing `Location`? 

Thanks again for fixing this!

**Review by kuhe — APPROVED** [review]

**Comment by wallind** [comment]
HUZZZAHHHHHHH
krunk
!

**Pull request merged**